include/RAJA/policy/sycl/forall.hpp

```diff
@@ -0,0 +1,189 @@
+ *
+ *          These methods should work on any platform that supports SYCL.
+ *
+ ******************************************************************************
+ */
+
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-19, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef RAJA_forall_sycl_HPP
+#define RAJA_forall_sycl_HPP
+
+#include "RAJA/config.hpp"
+
+#if defined(RAJA_ENABLE_SYCL)
+
+#include "CL/sycl.hpp"
+
+#include "RAJA/util/types.hpp"
+
+#include "RAJA/policy/sycl/policy.hpp"
+
+#include "RAJA/index/IndexSet.hpp"
+#include "RAJA/index/ListSegment.hpp"
+#include "RAJA/index/RangeSegment.hpp"
+
+#include "RAJA/internal/fault_tolerance.hpp"
+
+#include "RAJA/pattern/forall.hpp"
+
+
+namespace RAJA
+{
+
+namespace policy
+{
+namespace sycl
+{
+
+namespace impl
+{
+
+/*!
+ ******************************************************************************
+ *
+ * \brief calculate gridDim from length of iteration and blockDim
+ *
+ ******************************************************************************
+ */
+RAJA_INLINE
+sycl_dim_t getGridDim(size_t len, sycl_dim_t block_size)
+{
+  sycl_dim_t gridSize{block_size * ((len + block_size - 1) / block_size)};
+
+  return gridSize;
+}
+
+}  // namespace impl
+
+
+//
+////////////////////////////////////////////////////////////////////////
+//
+// Function templates for SYCL execution over iterables.
+//
+////////////////////////////////////////////////////////////////////////
+//
+
+template <typename Iterable, typename LoopBody, size_t BlockSize, bool Async>
+RAJA_INLINE void forall_impl(sycl_exec<BlockSize, Async>,
+                             Iterable&& iter,
+                             LoopBody&& loop_body)
+{
+
+  using Iterator  = camp::decay<decltype(std::begin(iter))>;
+  using LOOP_BODY = camp::decay<LoopBody>;
+  using IndexType = camp::decay<decltype(std::distance(std::begin(iter), std::end(iter)))>;
+
+  //
+  // Compute the requested iteration space size
+  //
+  Iterator begin = std::begin(iter);
+  Iterator end = std::end(iter);
+  IndexType len = std::distance(begin, end);
+  IndexType offset = *begin;
+
+  // Only launch kernel if we have something to iterate over
+  if (len > 0 && BlockSize > 0) {
+
+    //
+    // Compute the number of blocks
+    //
+    sycl_dim_t blockSize{BlockSize};
+    sycl_dim_t gridSize = impl::getGridDim(static_cast<size_t>(len), BlockSize);
+
+    cl::sycl::queue q = ::RAJA::sycl::detail::getQueue();
+
+    // Non-trivially copyable
+    Iterator* idx = (Iterator*) cl::sycl::malloc_device(sizeof(Iterator), q);
+    auto e = q.memcpy(idx, &begin, sizeof(Iterator));
+    e.wait();
+
+    LOOP_BODY* lbody = (LOOP_BODY*) cl::sycl::malloc_device(sizeof(LOOP_BODY), q);
```

Comment:
Are these mallocs pretty fast? I would be interested in knowing if they are negligible compared to the kernel run-time.

Reply:
I've started to implement the RAJA Perf Suite using the Sycl policies.  Getting the loop_body and iterator to the device are showing overheads.  I'm updating this to use unique_ptr's which can be passed to buffer allowing the for the runtime to take ownership and not block to free.  This is still showing overhead and I'm looking into it.

The Kernel policies I'm working on are showing very little overhead.